pkg/driver/vol_statter.go

```diff
@@ -0,0 +1,175 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package driver
+
+import (
+	"github.com/container-storage-interface/spec/lib/go/csi"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	"k8s.io/kubernetes/pkg/volume/util/fs"
+	"sync"
+	"time"
+)
+
+type volMetrics struct {
+	volPath   string
+	timeStamp time.Time
+	volUsage  []*csi.VolumeUsage
+}
+
+var (
+	volUsageCache        = make(map[string]*volMetrics)
+	volStatterJobTracker = make(map[string]bool)
+	fsRateLimiter        = make(map[string]int)
+	mu                   sync.RWMutex
+	jitter               = time.Duration(5 * time.Minute)
+)
+
+type VolStatter interface {
+	computeVolumeMetrics(volId, volPath string, refreshRate float64, fsRateLimit int) (*volMetrics, error)
+	retrieveFromCache(volId string) (*volMetrics, bool)
+	removeFromCache(volId string)
+}
+
+type VolStatterImpl struct {
+}
+
+func NewVolStatter() VolStatter {
+	return &VolStatterImpl{}
+}
+
+func (v VolStatterImpl) computeVolumeMetrics(volId, volPath string, refreshRate float64, fsRateLimit int) (*volMetrics, error) {
+	if value, ok := v.retrieveFromCache(volId); ok {
```

Comment:
What if I have two PVs mounted on this node with the same volume ID but different target path. Say volume A fs-123:/a and volume B fs-123:/b. It will be random which metric I get back.

Reply:
Maybe not random, but one of A or B will be reported for both A and B. 

However if you index the results cache by volume ID + path but the tracker just by volume ID, there might be situations where the metric for B is never computed. Let's say kubelet always calls get stats on A before B, then B will never get computed since the computation for A was just started and is in progress...

Reply:
IIRC, volumeId from above consists of file system ID + subpath. So `fs-123:/a` and `fs-123:/b` will be two different volumes and metric will be computed for both.

The cache will ensure that a metric for volumeId `fs-123:/a` mounted separately as two volumes will not be computed.

Reply:
ok, makes sense, i see now you distinguish between `fsid` for fs rate limit and `volid` for metrics cache